### PR TITLE
migrate module "l10n_es_devolucion_remesas" from 6.1 to "account_payment...

### PR DESCRIPTION
**Pull request by OSguard**
Add module to handle payment return and reopen the related invoice. Code was migrated from the l10n_es_devolucion_remesas module for version 6.1

### CODE REVIEW

File: account_payment_return/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    Provides a object to handle payment returns
+    """,
+    'data': [
+            'security/ir.model.access.csv',
+            'security/account_payment_return_security.xml',
+            'payment_return_view.xml',
+            'payment_return_data.xml',
+            'account_invoice_view.xml',
+    ],
+    'demo_xml': [],
```
Comment (bwrsandman) [inline account_payment_return/__openerp__.py]: This key is now named simply `'demo'`

File: account_payment_return/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    'installable': True,
+    'active': False,
```
Comment (bwrsandman) [inline account_payment_return/__openerp__.py]: This key can be removed.

File: account_payment_return/payment_return.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                                              'cancelled':[('readonly',True)]
+                                              }),
+        'date' : fields.date('Return date',
+                             help="This date will be used as the date for the "
+                             "account entry.",
+                             states={
+                                     'done':[('readonly',True)],
+                                     'cancelled':[('readonly',True)]
+                                     }),
+        'name' : fields.char("Reference", size=64, required=True,
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Unless you have a specific reason to set a hard limit on the size, you should ommit the `'size'` paramatter.

File: account_payment_return/payment_return.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                        ('res_id', '=', id),
+                        ('res_model', '=', 'payment.return'),
+                    ], context=context)
+            if id_at:
+                attachment_obj.unlink(cr, uid, id_at, context)
+        
+        return super(payment_order_return, self).unlink(cr, uid, ids,
+                                                        context=context)
+
+    def action_confirm(self, cr, uid, ids, *args):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: `*args` should be expanded to set parametters. At least have context out here.

File: account_payment_return/payment_return.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        invoice_obj = self.pool.get('account.invoice')
+        move_obj = self.pool.get('account.move')
+        move_line_obj = self.pool.get('account.move.line')
+        reconcile_obj = self.pool.get('account.move.reconcile')
+        return_line_obj = self.pool.get('payment.return.line')
+        for payment_return in self.browse(cr, uid, ids):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+            }
+            # Period
+            if payment_return.period_id:
+                move['period_id'] = payment_return.period_id.id
+            else:
+                move['period_id'] = self.pool.get('account.period').find(cr,
+                    uid, dt=payment_return.date, 
+                    context={'company_id': payment_return.company_id.id})[0]
+
+            move_id = move_obj.create(cr, uid, move)
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                         'credit': 0,
+                         })
+                    lines2reconcile.append(move_line_id)
+                    bank_move_line_id = move_line_obj.copy(
+                        cr, uid, move_line_id,
+                        {
+                         'debit': 0,
+                         'credit': return_line.amount,
+                         'account_id': move_line.move_id.journal_id.default_credit_account_id.id,
+                         })
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missin context propagation

File: account_payment_return/payment_return.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+                # Break old reconcile and make a new one with at least three moves
+                reconcile_obj.unlink(cr, uid, [old_reconcile.id])
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+                move_line_obj.reconcile_partial(cr, uid, lines2reconcile)
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+                move_line = move_line_obj.browse(cr, uid, move_line_id)
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+                return_line_obj.write(cr, uid, return_line.id, 
+                    {'reconcile_id': move_line.reconcile_partial_id.id})
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+                # Mark invoice as payment refused
+                invoice_obj.write(cr, uid, [invoice.id],
+                                  {'payment_returned': True})
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+            move_obj.button_validate(cr, uid, [move_id])
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+            self.write(cr, uid, payment_return.id,
+                       {'state': 'done', 'move_id': move_id})
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+        return True
+
+    def action_cancel(self, cr, uid, ids, *args):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: `'*args'` can be expanded to set parametters

File: account_payment_return/payment_return.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+        invoice_obj = self.pool.get('account.invoice')
+        move_obj = self.pool.get('account.move')
+        move_line_obj = self.pool.get('account.move.line')
+        reconcile_obj = self.pool.get('account.move.reconcile')
+        return_line_obj = self.pool.get('payment.return.line')
+        for payment_return in self.browse(cr, uid, ids):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+            if payment_return.move_id:
+                for return_line in payment_return.lines_id:
+                    if return_line.reconcile_id:
+                        reconcile = return_line.reconcile_id
+                        lines2reconcile = [x.id for x in
+                                           reconcile.line_partial_ids if 
+                                           x.move_id<>payment_return.move_id]
+                        reconcile_obj.unlink(cr, uid, [reconcile.id])
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+                        move_line_obj.reconcile(cr, uid, lines2reconcile)
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+                        return_line_obj.write(cr, uid, return_line.id, 
+                                              {'reconcile_id': False}
+                                              )
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py:222]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+                    # Remove payment refused flag on invoice
+                    invoice_obj.write(cr, uid, [return_line.invoice_id.id],
+                                      {'payment_returned': False})
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+                move_obj.button_cancel(cr, uid, [payment_return.move_id.id])
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+                move_obj.unlink(cr, uid, [payment_return.move_id.id])
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+        self.write(cr, uid, ids, {'state': 'cancelled', 'move_id': False})
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+        return True
+
+    def action_draft(self, cr, uid, ids, *args):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: `'*args'` can be expanded

File: account_payment_return/payment_return.py [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+        self.write(cr, uid, ids, {'state': 'draft'})
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Missing context propagation

File: account_payment_return/payment_return.py [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from osv import orm, fields
+from tools.translate import _
+import decimal_precision as dp
+import time
+
+class payment_order_return(orm.Model):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Please match the class name to the model and filename (`payment_return`)

File: account_payment_return/payment_return.py [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+                move_obj.unlink(cr, uid, [payment_return.move_id.id])
+        self.write(cr, uid, ids, {'state': 'cancelled', 'move_id': False})
+        return True
+
+    def action_draft(self, cr, uid, ids, *args):
+        self.write(cr, uid, ids, {'state': 'draft'})
+        return True
+
+
+class payment_order_return_line(orm.Model):
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: - This should be in its own python file.
- Class name should match filename and model name (`payment_return_line`)

File: account_payment_return/payment_return.py [hunk 28]
```diff
@@ -0,0 +1,251 @@
...
+    _name = "payment.return.line"
+    _description = 'Payment return lines'
+    
+    _columns = {
+        'return_id' : fields.many2one('payment.return', 'Payment return', 
+            required=True, ondelete='cascade'),
+        'concept': fields.char('Concept', size=100,
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Unless there is a good reason to fix a hard limit on the text size, the `size` parametter can be ommited

File: account_payment_return/payment_return.py [hunk 29]
```diff
@@ -0,0 +1,251 @@
...
+            help="Readed from imported file. Only for reference."),
+        'reason' : fields.char('Return reason', size=100,
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: Unless there is a good reason to fix a hard limit on the text size, the `size` parametter can be ommited

File: account_payment_return/payment_return.py [hunk 30]
```diff
@@ -0,0 +1,251 @@
...
+            required=True, ondelete='cascade'),
+        'concept': fields.char('Concept', size=100,
+            help="Readed from imported file. Only for reference."),
+        'reason' : fields.char('Return reason', size=100, 
+            help="Readed from imported file. Only for reference."),
+        'invoice_id': fields.many2one('account.invoice', 'Associated invoice'),
+        'date' : fields.date('Return date',
+            help="Readed from imported file. Only for reference."),
+        'notes' : fields.text('Notes'),
+        'partner_name': fields.char('Partner name', size=100,
```
Comment (bwrsandman) [inline account_payment_return/payment_return.py]: - Unless there is a good reason to fix a hard limit on the text size, the `size` parametter can be ommited
- Is the partner name related to `partner_id`?
  - If so, why not have a field.related?

File: account_payment_return/payment_return_data.xml [hunk 31]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bwrsandman) [inline account_payment_return/payment_return_data.xml]: Filename should match model name (`ir_sequence_type_data.xml`, `ir_sequence_data.xml`, `mail_message_subtype_data.xml`)

**Comment by bwrsandman** [comment]
Hi. Thank you for your contribution.
I have left a few comments in the diff, though the overall quality of this PR is very good.

File: account_payment_return/payment_return.py (hunk 30 above)
Comment (OSguard) [inline account_payment_return/payment_return.py]: partner name come from the bank import can be different from partner_id.name in odoo

**Comment by OSguard** [comment]
i have some correction done and make it flake8 compatible (almost)

**Comment by bwrsandman** [comment]
:+1:

File: account_payment_return/mail_message_subtype_data.xml [hunk 32]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data noupdate="1">
+
+        <!-- Paypemt Return-related subtypes for messaging / Chatter -->
```
Comment (tafaRU) [inline account_payment_return/mail_message_subtype_data.xml]: I think Paypemt stands for Payment ;-) Am I right?

**Comment by OSguard** [comment]
yes you are right, fix it

**Comment by OSguard** [comment]
@tafaRU can you finish code review please

File: account_payment_return/payment_return.py [hunk 33]
```diff
@@ -0,0 +1,231 @@
...
+        'state': fields.selection([
+            ('draft', 'Draft'),
+            ('imported', 'Imported'),
+            ('done', 'Done'),
+            ('cancelled', 'Cancelled')
+            ], 'State', readonly=True),
+    }
+    _track = {
+        'state': {
+            'account_payment_return.mt_payment_return_state_change': lambda self, cr, uid, obj, ctx=None: True,
```
Comment (tafaRU) [inline account_payment_return/payment_return.py]: travis build says: E501 line too long (111 > 79 characters)

File: account_payment_return/payment_return.py [hunk 34]
```diff
@@ -0,0 +1,231 @@
...
+        },
+    }
+
+    _defaults = {
+        'company_id': lambda s, cr, uid, c: s.pool.get('res.company')._company_default_get(cr, uid, 'account', context=c),
```
Comment (tafaRU) [inline account_payment_return/payment_return.py]: travis build says: E501 line too long (122 > 79 characters)

File: account_payment_return/payment_return.py [hunk 35]
```diff
@@ -0,0 +1,231 @@
...
+        'state': 'draft',
+        'date': lambda *x: time.strftime('%Y-%m-%d %H:%M:%S'),
+        'name': lambda obj, cr, uid, context: obj.pool.get('ir.sequence').next_by_code(cr, uid, 'payment.return'),
```
Comment (tafaRU) [inline account_payment_return/payment_return.py]: travis build says: E501 line too long (114 > 79 characters)

File: account_payment_return/payment_return.py (hunk 34 above)
Comment (OSguard) [inline account_payment_return/payment_return.py]: i know but didn't find solution to fix this

Comment (tafaRU) [inline account_payment_return/payment_return.py]: https://github.com/initOS/account-payment/pull/1 should fix it

**Comment by OSguard** [comment]
THX a lot, something learned today, that we can break after ( in a function.
Now test is pass \o/

**Comment by tafaRU** [comment]
@OSguard ;-)
LGTM but I think the final decision goes to @pedrobaeza cosidering that, if I'm not wrong, he's the author of the original module.

**Comment by legalsylvain** [comment]
@pedrobaeza could you take a look on this old PR where you are mentioned ?

**Comment by max3903** [comment]
:+1:

**Comment by OSguard** [comment]
now it can be merged! @pedrobaeza

**Comment by pedrobaeza** [comment]
:+1:

**Pull request merged**